### PR TITLE
[Release] Hero 1.4.0 with Swift 4.2 support

### PR DESCRIPTION
**Pull request by SD10**
### Summary
- Adds support for Swift 4.2
- The example target for iOS does not compile because [CollectionKit](https://github.com/SoySauceLab/CollectionKit) needs a Swift 4.2 update still

### Related
Replaces #490 

### TODO ⏰ 
N/A

### CODE REVIEW

File: Sources/Debug Plugin/HeroDebugView.swift [hunk 1]
```diff
@@ -69,18 +69,18 @@ class HeroDebugView: UIView {
     addSubview(backgroundView)
 
     doneButton = UIButton(type: .system)
-    doneButton.setTitle("Done", for: .normal)
+    doneButton.setTitle("Done", for: UIControl.State.normal)
```
Comment (freak4pc) [inline Sources/Debug Plugin/HeroDebugView.swift]: You can just do `.normal`, no need to be this explicit :)

File: Sources/Debug Plugin/HeroDebugView.swift [hunk 2]
```diff
@@ -69,18 +69,18 @@ class HeroDebugView: UIView {
...
     doneButton.addTarget(self, action: #selector(onDone), for: .touchUpInside)
     backgroundView.addSubview(doneButton)
 
     perspectiveButton = UIButton(type: .system)
-    perspectiveButton.setTitle("3D View", for: .normal)
+    perspectiveButton.setTitle("3D View", for: UIControl.State.normal)
```
Comment (freak4pc) [inline Sources/Debug Plugin/HeroDebugView.swift]: same

File: Sources/Debug Plugin/HeroDebugView.swift [hunk 3]
```diff
@@ -69,18 +69,18 @@ class HeroDebugView: UIView {
...
     perspectiveButton.addTarget(self, action: #selector(onPerspective), for: .touchUpInside)
     backgroundView.addSubview(perspectiveButton)
 
     if showCurveButton {
       arcCurveButton = UIButton(type: .system)
-      arcCurveButton!.setTitle("Show Arcs", for: .normal)
+      arcCurveButton!.setTitle("Show Arcs", for: UIControl.State.normal)
```
Comment (freak4pc) [inline Sources/Debug Plugin/HeroDebugView.swift]: same

File: Sources/Debug Plugin/HeroDebugView.swift [hunk 4]
```diff
@@ -162,7 +162,6 @@ class HeroDebugView: UIView {
       startLocation = pinchGR.location(in: nil)
       startTranslation = translation
       startScale = scale
-      fallthrough
```
Comment (freak4pc) [inline Sources/Debug Plugin/HeroDebugView.swift]: Why did you remove `fallthrough` here? I think this needs to fall to the case below

File: Sources/Transition/HeroProgressRunner.swift [hunk 5]
```diff
@@ -61,12 +61,12 @@ class HeroProgressRunner {
     self.isReversed = reverse
     self.duration = totalTime
     displayLink = CADisplayLink(target: self, selector: #selector(displayUpdate(_:)))
-    displayLink!.add(to: RunLoop.main, forMode: RunLoopMode(rawValue: RunLoopMode.commonModes.rawValue))
+    displayLink!.add(to: RunLoop.main, forMode: RunLoop.Mode(rawValue: RunLoop.Mode.common.rawValue))
```
Comment (freak4pc) [inline Sources/Transition/HeroProgressRunner.swift]: This can simply be 
```swift
displayLink!.add(to: .main, forMode: .common)
```

File: Sources/Transition/HeroProgressRunner.swift [hunk 6]
```diff
@@ -61,12 +61,12 @@ class HeroProgressRunner {
...
   }
 
   func stop() {
     displayLink?.isPaused = true
-    displayLink?.remove(from: RunLoop.main, forMode: RunLoopMode(rawValue: RunLoopMode.commonModes.rawValue))
+    displayLink?.remove(from: RunLoop.main, forMode: RunLoop.Mode(rawValue: RunLoop.Mode.common.rawValue))
```
Comment (freak4pc) [inline Sources/Transition/HeroProgressRunner.swift]: same

**Comment by freak4pc** [comment]
By the way just realized one more issue here - won't you be breaking support for 4.0 this way? Maybe we should add conditional build statements?

**Comment by SD10** [comment]
Yeah, @freak4pc conditionals isn't a bad improvement 👍 Given that this library will move slow I'm almost tempted to make 1.3.1 the last Swift 4.0 release but I'll see what I can do

**Comment by freak4pc** [comment]
Even so, migrations to Swift 4.2 will take some time, so it's great supporting the last 2 at least.

**Comment by skydivedan** [comment]
Is backward compatibility a problem, though? 
- Version 1.3.x of Hero could be for Swift 4.0, 4.1 users. 
- Version 1.4.x could be for Swift 4.2 users.

Also, you could set `s.swift_version = 4.2` in the Podspec file, as an extra hint for Cocoapods. 

If there's a critical change that you think is needed for Swift 4 and Swift 4.2, you could rev 2 separate pods. Probably not ideal I guess, though. Just my $0.02 from an early adopter of Swift 4.2

**Comment by SD10** [comment]
@skydivedan 

> Is backward compatibility a problem, though?

I'm not really sure how much of an issue backward compatibility is and how long we would want to support both versions if we did.

I'm personally fine using Hero 1.3.1 and I can't think of any major upcoming Hero releases that would take place before most people have migrated to Swift 4.2. Unless there are strong objections I'll probably just end up pushing this

> If there's a critical change that you think is needed for Swift 4 and Swift 4.2, you could rev 2 separate pods. Probably not ideal I guess, though. Just my $0.02 from an early adopter of Swift 4.2

Yeah, probably not ideal 😅  But thanks for the input 👍

**Comment by 4oby** [comment]
@SD10 just FYI https://github.com/SoySauceLab/CollectionKit/releases/tag/2.2.0

**Comment by alexanderkhitev** [comment]
@SD10 any news?

**Comment by SD10** [comment]
Sorry for the delay everyone, I had finally burned out 😓 I feel like I'm slowly recovering but that was rough for me. 

I added backward compatibility for previous Swift versions and will release this now. Feel free to open an issue if I botched anything in the environment

**Comment by metabren** [comment]
No problem at all @SD10 — never a nice thing to go through. Thank you for this, and all the best for your recovery. Hope it's a speedy one!

**Comment by freak4pc** [comment]
Hey @SD10, 
Please take more time - I know hard this can be very up close but the community is here to support you. Feel free to at-me in every PR on any organization you need help with. 

Take care of yourself, the rest is meaningless ❤️
